flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/resource/UserResourceManager.java

```diff
@@ -0,0 +1,195 @@
+    }
+
+    private void checkResource(ResourceType resourceType, Path path) {
+        try {
+            FileSystem fs = path.getFileSystem();
+            // check resource exists firstly
+            if (!fs.exists(path)) {
+                throw new IllegalArgumentException(String.format("Resource [%s] not found.", path));
+            }
+
+            if (ResourceType.JAR.equals(resourceType)) {
+                // register directory is not allowed for jar resource
+                if (fs.getFileStatus(path).isDir()) {
+                    throw new IllegalArgumentException(
+                            String.format(
+                                    "Directory [%s] is not allowed for registering jar resource.",
+                                    path));
+                }
+
+                // Jar resource must ends with '.jar' suffix
+                if (!path.getName().toLowerCase().endsWith(".jar")) {
+                    throw new IllegalArgumentException(
+                            String.format(
+                                    "Registering jar resource [%s] must ends with '.jar' suffix.",
+                                    path));
+                }
+            }
+        } catch (IOException e) {
+            throw new FlinkRuntimeException(
+                    String.format("Failed to check resource [%s].", path), e);
+        }
+    }
+
+    @VisibleForTesting
+    protected URL downloadResource(Path remotePath) {
+        // get local resource path
+        java.nio.file.Path path = Paths.get(localTmpDir, remotePath.getName()).toAbsolutePath();
+        Path localResourcePath = new Path(path.toUri());
+        try {
+            FileUtils.copy(remotePath, localResourcePath, true);
+            LOG.info(
+                    "Download resource [{}] to local path [{}] successfully.",
+                    remotePath,
+                    localResourcePath);
+        } catch (IOException e) {
+            throw new FlinkRuntimeException(
+                    String.format(
+                            "Failed to download resource [%s] to local path [%s].",
+                            remotePath, localTmpDir),
+                    e);
+        }
+        return getURLFromPath(localResourcePath);
+    }
+
+    @VisibleForTesting
+    protected URL getURLFromPath(Path path) {
+        try {
+            // if scheme is null, rewrite it to file
+            if (path.toUri().getScheme() == null) {
+                path = path.makeQualified(FileSystem.getLocalFileSystem());
+            }
+            return path.toUri().toURL();
+        } catch (MalformedURLException e) {
+            throw new FlinkRuntimeException(
+                    String.format("Failed to convert the resource path [%s] to URL.", path), e);
+        }
+    }
+
+    private void checkJarFile(URL jarUrl) {
+        try {
+            // verify that we can open the Jar file
+            new JarFile(new File(jarUrl.toURI()));
```

Comment:
Please use `org.apache.flink.util.JarUtils#checkJarFile` to check jar which provides more readable exception messages.